openedx/core/djangoapps/user_api/accounts/tests/test_views.py

```diff
@@ -245,6 +254,92 @@ def test_get_username_not_logged_in(self):
+        assert response.status_code == status.HTTP_400_BAD_REQUEST
+        assert response.data == {'message': 'retirement_id must be specified.'}
+
+    def test_cancel_retirement_does_not_exist(self):
+        """
+        Test that cancel_retirement throws 400 if no retirement status exists.
+        """
+        client = self.login_client('staff_client', 'staff_user')
+        url = reverse("cancel_account_retirement")
+        response = client.post(url, data={'retirement_id': 1})
+        assert response.status_code == status.HTTP_400_BAD_REQUEST
+        assert response.data == {"message": 'Retirement does not exist!'}
+
+    def test_cancel_retirement_not_pending(self):
+        """
+        Test that cancel_retirement throws 400 if retirement state is not PENDING.
+        """
+        client = self.login_client('staff_client', 'staff_user')
+        retirement_state = RetirementStateFactory.create(state_name='NOT_PENDING', state_execution_order=1)
+        user_retirement_status = UserRetirementStatusFactory.create(
+            user=self.user,
+            current_state=retirement_state,
+            last_state=retirement_state,
+            original_email=self.user.email,
+            created=datetime.datetime.now(pytz.UTC)
+        )
+        url = reverse("cancel_account_retirement")
+        response = client.post(url, data={'retirement_id': user_retirement_status.id})
+        assert response.status_code == status.HTTP_400_BAD_REQUEST
+        assert response.data == {
+            "message": f"Retirement requests can only be cancelled for users in the PENDING state. "
+                       f"Current request state for '{user_retirement_status.original_username}': "
+                       f"{user_retirement_status.current_state.state_name}"
+        }
+
+    def test_cancel_retirement_successful(self):
+        """
+        Test that cancel_retirement does the following things properly:
+        1. Restore user's email
+        2. Reset user's password
+        3. Delete Retirement Status entry
+        """
+        client = self.login_client('staff_client', 'staff_user')
+        retirement_state = RetirementStateFactory.create(state_name='PENDING', state_execution_order=1)
+        user_retirement_status = UserRetirementStatusFactory.create(
+            user=self.user,
+            current_state=retirement_state,
+            last_state=retirement_state,
+            original_email=self.user.email,
+            created=datetime.datetime.now(pytz.UTC)
+        )
+        assert UserRetirementStatus.objects.count() == 1
+
+        url = reverse("cancel_account_retirement")
+        response = client.post(url, data={'retirement_id': user_retirement_status.id})
+        assert response.status_code == status.HTTP_200_OK
+        assert response.data == {"success": True}
+        assert user_retirement_status.user.email == user_retirement_status.original_email
+        assert user_retirement_status.user.has_usable_password()
```

Comment:
Get the user associated with retirement status, refresh from DB after API call, and do assertion checks on that user for the values.

Reply:
Did a similar thing to verify user has a useable password now.